File: frontend/app/lib/simulator/emergent/searcherAgent.js
/**
 * SearcherAgent - Individual searcher with position and behavior
 *
 * Each searcher is an actual agent that:
 * - Has a position (lat/lng)
 * - Moves through the environment
 * - Can detect pets within their detection range
 * - Follows search patterns (grid, spiral, probability-based)
 *
 * This is TRUE emergent simulation - detection happens when
 * searcher and pet are in proximity, not from probability math.
 */

// Detection range in miles (how far can a searcher see/hear a pet)
const BASE_DETECTION_RANGE = {
  DAY_VISUAL: 0.05,    // ~80 meters line of sight
  DAY_AUDIO: 0.15,     // ~240 meters if pet makes noise
  NIGHT_VISUAL: 0.01,  // ~16 meters with flashlight
  NIGHT_AUDIO: 0.20,   // Better at night (quieter)
};

// Searcher speed in miles per minute
const SEARCHER_SPEED = {
  WALKING: 0.05,       // 3 mph
  SLOW_SEARCH: 0.025,  // 1.5 mph (careful searching)
  RUNNING: 0.10,       // 6 mph (responding to sighting)
};

/**
 * SearcherAgent class
 */
export class SearcherAgent {
  constructor(id, config, homePosition, random) {
    this.id = id;
    this.isOwner = id === 0;  // First searcher is always the owner
    this.random = random;

    // Position
    this.lat = homePosition.lat;
    this.lng = homePosition.lng;
    this.homeLat = homePosition.lat;
    this.homeLng = homePosition.lng;

    // Search config
    this.searchRadius = config.searchRadiusMiles || 2.0;
    this.strategy = config.searchStrategy || 'PROBABILITY';

    // State
    this.isActive = false;
    this.currentSpeed = SEARCHER_SPEED.WALKING;
    this.heading = this.random() * 360;  // Initial random heading

    // For spiral/grid patterns
    this.spiralRadius = 0.05;  // Start close to home
    this.spiralAngle = (id * 137.5) % 360;  // Golden angle distribution
    this.gridX = 0;
    this.gridY = 0;

    // For probability-based search
    this.targetLat = null;
    this.targetLng = null;

    // Track distance covered
    this.distanceCovered = 0;

    // Recall training (owner knows pet's name, habits)
    this.recallBonus = this.isOwner ? 0.3 : 0;
  }

  /**
   * Activate searcher (start searching)
   */
  activate() {
    this.isActive = true;
  }

  /**
   * Move searcher for one time step
   */
  move(timeStepMinutes, focusLocation = null) {
    if (!this.isActive) return;

    // If there's a focus location (reported sighting), head there
    if (focusLocation && this.strategy !== 'GRID') {
      this.moveTowardTarget(focusLocation.lat, focusLocation.lng, timeStepMinutes);
      return;
    }

    // Otherwise, follow search pattern
    switch (this.strategy) {
      case 'GRID':
        this.moveGrid(timeStepMinutes);
        break;
      case 'SPIRAL':
        this.moveSpiral(timeStepMinutes);
        break;
      case 'PROBABILITY':
        this.moveProbability(timeStepMinutes);
        break;
      case 'RANDOM':
      default:
        this.moveRandom(timeStepMinutes);
    }
  }

  /**
   * Grid search pattern - systematic coverage
   */
  moveGrid(timeStepMinutes) {
    const gridSpacing = 0.1;  // Miles between grid lines
    const distance = this.currentSpeed * timeStepMinutes;

    // Move along current heading
    const radHeading = this.heading * Math.PI / 180;
    const dLat = distance * Math.cos(radHeading) / 69;  // 69 miles per degree lat
    const dLng = distance * Math.sin(radHeading) / (69 * Math.cos(this.lat * Math.PI / 180));

    this.lat += dLat;
    this.lng += dLng;
    this.distanceCovered += distance;

    // Check if we've gone too far from home
    const distFromHome = this.distanceFrom(this.homeLat, this.homeLng);
    if (distFromHome > this.searchRadius) {
      // Turn around
      this.heading = (this.heading + 180) % 360;
      // Shift to next grid line
      this.gridX += 1;
      if (this.gridX > this.searchRadius / gridSpacing) {
        this.gridX = 0;
        this.gridY += 1;
      }
    }
  }

  /**
   * Spiral search pattern - expanding circles
   */
  moveSpiral(timeStepMinutes) {
    const distance = this.currentSpeed * timeStepMinutes;

    // Expand the spiral
    this.spiralAngle += 15;  // Degrees per step
    this.spiralRadius += 0.002;  // Expand outward

    // Cap at search radius
    if (this.spiralRadius > this.searchRadius) {
      this.spiralRadius = 0.05;  // Reset to start
    }

    // Calculate new position
    const radAngle = this.spiralAngle * Math.PI / 180;
    this.lat = this.homeLat + (this.spiralRadius / 69) * Math.cos(radAngle);
    this.lng = this.homeLng + (this.spiralRadius / (69 * Math.cos(this.homeLat * Math.PI / 180))) * Math.sin(radAngle);

    this.distanceCovered += distance;
  }

  /**
   * Probability-based search - focus on likely areas
   */
  moveProbability(timeStepMinutes) {
    const distance = this.currentSpeed * timeStepMinutes;

    // If no target or reached target, pick a new one
    if (!this.targetLat || this.distanceFrom(this.targetLat, this.targetLng) < 0.02) {
      // Focus on areas where pets are likely to be
      // (close to home, with cover)
      const angle = this.random() * 360;
      const radius = Math.pow(this.random(), 2) * this.searchRadius;  // Bias toward closer areas

      const radAngle = angle * Math.PI / 180;
      this.targetLat = this.homeLat + (radius / 69) * Math.cos(radAngle);
      this.targetLng = this.homeLng + (radius / (69 * Math.cos(this.homeLat * Math.PI / 180))) * Math.sin(radAngle);
    }

    this.moveTowardTarget(this.targetLat, this.targetLng, timeStepMinutes);
  }

  /**
   * Random walk (uncoordinated search)
   */
  moveRandom(timeStepMinutes) {
    const distance = this.currentSpeed * timeStepMinutes;

    // Random heading changes
    this.heading += (this.random() - 0.5) * 60;  // ±30 degrees

    const radHeading = this.heading * Math.PI / 180;
    const dLat = distance * Math.cos(radHeading) / 69;
    const dLng = distance * Math.sin(radHeading) / (69 * Math.cos(this.lat * Math.PI / 180));

    this.lat += dLat;
    this.lng += dLng;
    this.distanceCovered += distance;

    // Bounce back if too far from home
    const distFromHome = this.distanceFrom(this.homeLat, this.homeLng);
    if (distFromHome > this.searchRadius) {
      // Head back toward home
      this.heading = this.headingTo(this.homeLat, this.homeLng);
    }
  }

  /**
   * Move toward a target position
   */
  moveTowardTarget(targetLat, targetLng, timeStepMinutes) {
    const distance = this.currentSpeed * timeStepMinutes;

    // Calculate heading to target
    this.heading = this.headingTo(targetLat, targetLng);

    const radHeading = this.heading * Math.PI / 180;
    const dLat = distance * Math.cos(radHeading) / 69;
    const dLng = distance * Math.sin(radHeading) / (69 * Math.cos(this.lat * Math.PI / 180));

    this.lat += dLat;
    this.lng += dLng;
    this.distanceCovered += distance;
  }

  /**
   * Check if this searcher can detect a pet at the given position
   *
   * Returns: { detected: boolean, distance: number, method: string }
   */
  checkDetection(petLat, petLng, petState, currentHour, environment) {
    if (!this.isActive) return { detected: false };

    const distance = this.distanceFrom(petLat, petLng);

    // Determine detection range based on conditions
    const isNight = currentHour < 6 || currentHour > 20;
    const isHiding = petState === 'HIDING';
    const isFleeing = petState === 'FLEEING';

    // Visual detection range
    let visualRange = isNight ? BASE_DETECTION_RANGE.NIGHT_VISUAL : BASE_DETECTION_RANGE.DAY_VISUAL;

    // Concealment reduces visual range
    const concealment = environment.getConcealmentAt(petLat, petLng);
    visualRange *= (1 - concealment * 0.7);

    // Hiding pets are harder to see
    if (isHiding) {
      visualRange *= 0.3;
    }

    // Audio detection range
    let audioRange = isNight ? BASE_DETECTION_RANGE.NIGHT_AUDIO : BASE_DETECTION_RANGE.DAY_AUDIO;

    // Fleeing pets make more noise
    if (isFleeing) {
      audioRange *= 1.5;
    }

    // Hiding pets are quiet
    if (isHiding) {
      audioRange *= 0.2;
    }

    // Owner bonus (knows pet's sounds)
    if (this.isOwner) {
      audioRange *= 1.3;
    }

    // Check detection
    if (distance <= visualRange) {
      return { detected: true, distance, method: 'visual' };
    }

    if (distance <= audioRange) {
      // Audio detection has random chance
      const audioDetectProb = 0.3 * (1 - distance / audioRange);
      if (this.random() < audioDetectProb) {
        return { detected: true, distance, method: 'audio' };
      }
    }

    return { detected: false, distance };
  }

  /**
   * Calculate distance to a point
   */
  distanceFrom(lat, lng) {
    const R = 3959;  // Earth radius in miles
    const dLat = (lat - this.lat) * Math.PI / 180;
    const dLng = (lng - this.lng) * Math.PI / 180;
    const a = Math.sin(dLat / 2) ** 2 +
              Math.cos(this.lat * Math.PI / 180) * Math.cos(lat * Math.PI / 180) *
              Math.sin(dLng / 2) ** 2;
    const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
    return R * c;
  }

  /**
   * Calculate heading to a point
   */
  headingTo(lat, lng) {
    const dLng = (lng - this.lng) * Math.PI / 180;
    const y = Math.sin(dLng) * Math.cos(lat * Math.PI / 180);
    const x = Math.cos(this.lat * Math.PI / 180) * Math.sin(lat * Math.PI / 180) -
              Math.sin(this.lat * Math.PI / 180) * Math.cos(lat * Math.PI / 180) * Math.cos(dLng);
    return (Math.atan2(y, x) * 180 / Math.PI + 360) % 360;
  }

  /**
   * Get current position
   */
  getPosition() {
    return { lat: this.lat, lng: this.lng };
  }
}

/**
 * Create a team of searcher agents
 */
export function createSearchTeam(count, config, homePosition, random) {
  const searchers = [];
  for (let i = 0; i < count; i++) {
    searchers.push(new SearcherAgent(i, config, homePosition, random));
  }
  return searchers;
}

File: simulation/agents/searcher_agent.py
"""
Searcher Agent for Monte Carlo Simulation

Implements discrete searcher agents from BEHAVIORAL_PROFILES.md Part 11:
- Multiple searcher types (OWNER, VOLUNTEER, PROFESSIONAL, etc.)
- Search strategies (ProfileAware, SightingChaser, etc.)
- Fatigue and dedication modeling
"""

import math
import random
from typing import Optional, Any, Tuple, List, Dict
from dataclasses import dataclass, field
from enum import Enum

from ..core.types import (
    SearcherType, SearchStrategy, SightingReport
)
from ..core.constants import SEARCHER_TYPE_PARAMS
from ..core.utils import (
    distance, direction_to, offset_position, normalize_angle,
    get_time_period, clamp
)


@dataclass
class SearcherState:
    """Runtime state of a searcher."""
    position: Tuple[float, float]
    is_active: bool = False
    current_fatigue: float = 0.0
    hours_searched_today: float = 0.0
    total_hours_searched: float = 0.0
    cells_searched: List[Tuple[int, int]] = field(default_factory=list)
    current_target: Optional[Tuple[float, float]] = None
    heading: float = 0.0


class SearcherAgent:
    """
    Individual searcher agent that moves through the environment
    searching for the lost pet.

    Searchers have:
    - A type (owner, volunteer, professional, etc.)
    - A search strategy
    - Fatigue that accumulates and reduces effectiveness
    - Limited hours per day
    """

    def __init__(
        self,
        searcher_id: int,
        searcher_type: SearcherType,
        strategy: SearchStrategy,
        home_position: Tuple[float, float],
        search_radius_m: float,
        is_owner: bool = False,
        seed: Optional[int] = None
    ):
        self.id = searcher_id
        self.searcher_type = searcher_type
        self.strategy = strategy
        self.is_owner = is_owner
        self.home_position = home_position
        self.search_radius_m = search_radius_m
        self.rng = random.Random(seed)

        # Load type-specific parameters
        type_key = searcher_type.value.upper()
        self.params = SEARCHER_TYPE_PARAMS.get(type_key, SEARCHER_TYPE_PARAMS["VOLUNTEER"])

        # Initialize state
        self.state = SearcherState(position=home_position)
        self.state.heading = self.rng.uniform(0, 2 * math.pi)

        # Strategy-specific state
        self.spiral_angle = 0
        self.spiral_radius = 50  # Start 50m from center
        self.grid_x = 0
        self.grid_y = 0
        self.grid_spacing = 100  # 100m grid cells

        # Sighting tracking (for SightingChaser strategy)
        self.known_sightings: List[SightingReport] = []
        self.current_sighting_target: Optional[SightingReport] = None

    def activate(self, current_hour: int):
        """Activate the searcher if conditions are met."""
        # Check if within search hours (typically 7am - 9pm)
        if 7 <= current_hour <= 21:
            self.state.is_active = True

    def deactivate(self):
        """Deactivate the searcher."""
        self.state.is_active = False

    def tick(
        self,
        hours_delta: float,
        environment: Any,
        current_hour: int,
        sightings: List[SightingReport] = None
    ):
        """Execute one simulation tick for this searcher."""
        if not self.state.is_active:
            # Check if should activate
            self.activate(current_hour)
            if not self.state.is_active:
                return

        # Check if exceeded daily hours
        max_hours = self.params["hours_per_day"]
        if self.state.hours_searched_today >= max_hours:
            self.deactivate()
            return

        # Check time of day
        if current_hour < 7 or current_hour > 21:
            self.deactivate()
            return

        # Update fatigue
        fatigue_rate = 1 / (self.params["fatigue_resistance"] * 8)  # Hours to full fatigue
        self.state.current_fatigue = clamp(
            self.state.current_fatigue + fatigue_rate * hours_delta,
            0, 1
        )

        # Update sightings knowledge
        if sightings:
            self.known_sightings = sightings

        # Move according to strategy
        self.move(hours_delta, environment)

        # Update tracking
        self.state.hours_searched_today += hours_delta
        self.state.total_hours_searched += hours_delta

    def move(self, hours_delta: float, environment: Any):
        """Move the searcher according to their strategy."""
        # Calculate speed (affected by fatigue)
        base_speed = 4000  # 4 km/h walking speed in m/h
        fatigue_modifier = 1 - (self.state.current_fatigue * 0.4)
        speed = base_speed * fatigue_modifier

        distance_m = speed * hours_delta

        # Get direction based on strategy
        if self.strategy == SearchStrategy.NAIVE_EXPANDING:
            direction = self._naive_expanding_direction()
        elif self.strategy == SearchStrategy.PROFILE_AWARE:
            direction = self._profile_aware_direction(environment)
        elif self.strategy == SearchStrategy.SIGHTING_CHASER:
            direction = self._sighting_chaser_direction()
        elif self.strategy == SearchStrategy.COORDINATED_GRID:
            direction = self._coordinated_grid_direction()
        elif self.strategy == SearchStrategy.TRAP_FOCUSED:
            direction = self._trap_focused_direction(environment)
        else:
            direction = self._naive_expanding_direction()

        # Move
        new_position = offset_position(self.state.position, distance_m, direction, self.rng)

        # Check bounds
        dist_from_home = distance(new_position, self.home_position)
        if dist_from_home > self.search_radius_m:
            # Bounce back toward center
            direction = direction_to(self.state.position, self.home_position)
            new_position = offset_position(self.state.position, distance_m, direction, self.rng)

        self.state.position = new_position
        self.state.heading = direction

    def _naive_expanding_direction(self) -> float:
        """
        Naive expanding circle strategy.

        Start near escape point and spiral outward.
        Simple but not optimal.
        """
        # Expanding spiral from home
        self.spiral_angle += 0.1
        self.spiral_radius += 5  # Expand 5m per step

        # Cap at search radius
        if self.spiral_radius > self.search_radius_m:
            self.spiral_radius = 50
            self.spiral_angle = self.rng.uniform(0, 2 * math.pi)

        # Calculate target on spiral
        target_lat = self.home_position[0] + (self.spiral_radius / 111000) * math.cos(self.spiral_angle)
        target_lon = self.home_position[1] + (self.spiral_radius / (111000 * math.cos(math.radians(self.home_position[0])))) * math.sin(self.spiral_angle)

        return direction_to(self.state.position, (target_lat, target_lon))

    def _profile_aware_direction(self, environment: Any) -> float:
        """
        Profile-aware search strategy.

        Uses knowledge of pet behavior to search likely areas first:
        - Hiding spots for cats
        - Travel corridors for dogs
        - Water sources
        """
        if not self.state.current_target:
            # Pick a new target based on probability
            self.state.current_target = self._pick_profile_target(environment)

        if self.state.current_target:
            dist = distance(self.state.position, self.state.current_target)
            if dist < 20:  # Reached target
                self.state.current_target = None
                return self._naive_expanding_direction()

            return direction_to(self.state.position, self.state.current_target)

        return self._naive_expanding_direction()

    def _pick_profile_target(self, environment: Any) -> Optional[Tuple[float, float]]:
        """Pick a high-probability target location."""
        # Would query environment for hiding spots, water sources, etc.
        # For now, bias toward areas near home
        angle = self.rng.uniform(0, 2 * math.pi)
        # Bias toward closer areas
        radius = self.rng.triangular(50, 100, self.search_radius_m * 0.5)

        lat = self.home_position[0] + (radius / 111000) * math.cos(angle)
        lon = self.home_position[1] + (radius / (111000 * math.cos(math.radians(self.home_position[0])))) * math.sin(angle)

        return (lat, lon)

    def _sighting_chaser_direction(self) -> float:
        """
        Sighting chaser strategy.

        Prioritize investigating recent sightings.
        """
        if self.known_sightings:
            # Find most recent unvisited sighting
            if not self.current_sighting_target:
                # Sort by timestamp, most recent first
                sorted_sightings = sorted(
                    self.known_sightings,
                    key=lambda s: s.timestamp,
                    reverse=True
                )
                if sorted_sightings:
                    self.current_sighting_target = sorted_sightings[0]

            if self.current_sighting_target:
                dist = distance(self.state.position, self.current_sighting_target.location)
                if dist < 30:
                    # Investigated this sighting
                    self.current_sighting_target = None
                    return self._naive_expanding_direction()

                return direction_to(self.state.position, self.current_sighting_target.location)

        return self._naive_expanding_direction()

    def _coordinated_grid_direction(self) -> float:
        """
        Coordinated grid search strategy.

        Systematic coverage of the area in a grid pattern.
        Multiple searchers would coordinate to avoid overlap.
        """
        # Calculate grid target
        grid_x_offset = self.grid_x * self.grid_spacing
        grid_y_offset = self.grid_y * self.grid_spacing

        target_lat = self.home_position[0] + (grid_y_offset / 111000)
        target_lon = self.home_position[1] + (grid_x_offset / (111000 * math.cos(math.radians(self.home_position[0]))))

        dist = distance(self.state.position, (target_lat, target_lon))

        if dist < 20:
            # Move to next grid cell
            self.grid_x += 1
            if self.grid_x * self.grid_spacing > self.search_radius_m:
                self.grid_x = -int(self.search_radius_m / self.grid_spacing)
                self.grid_y += 1
            if self.grid_y * self.grid_spacing > self.search_radius_m:
                self.grid_y = -int(self.search_radius_m / self.grid_spacing)

        return direction_to(self.state.position, (target_lat, target_lon))

    def _trap_focused_direction(self, environment: Any) -> float:
        """
        Trap-focused strategy.

        Check trap locations regularly, otherwise patrol nearby.
        """
        # Would check trap locations from environment
        # For now, similar to profile-aware but with trap checks
        return self._profile_aware_direction(environment)

    def check_detection(
        self,
        pet_position: Tuple[float, float],
        pet_status: str,
        current_hour: int
    ) -> Tuple[bool, float]:
        """
        Check if this searcher detects the pet.

        Returns (detected, distance)
        """
        dist = distance(self.state.position, pet_position)

        # Get detection range (affected by fatigue)
        base_range = self.params["detection_range_m"]
        fatigue_modifier = 1 - (self.state.current_fatigue * 0.3)
        detection_range = base_range * fatigue_modifier

        # Night reduces range
        if current_hour < 7 or current_hour > 20:
            detection_range *= 0.3

        # Pet state affects visibility
        if pet_status == "hiding":
            detection_range *= 0.3
        elif pet_status == "fleeing":
            detection_range *= 1.2

        if dist <= detection_range:
            # Detection probability based on distance
            detection_prob = 1 - (dist / detection_range) * 0.5

            # Owner gets recall bonus
            if self.is_owner:
                detection_prob += self.params["recall_bonus"]

            if self.rng.random() < detection_prob:
                return True, dist

        return False, dist

    def attempt_capture(
        self,
        pet_agent: Any,
        distance_m: float
    ) -> bool:
        """
        Attempt to capture the detected pet.

        Returns True if capture successful.
        """
        from ..core.utils import calculate_capture_probability

        capture_prob = calculate_capture_probability(
            animal_state=pet_agent.state,
            animal_profile=pet_agent.profile,
            searcher_type=self.searcher_type.value.upper(),
            is_owner=self.is_owner,
            distance_m=distance_m,
            time_of_day=get_time_period(self.state.hours_searched_today % 24),
            rng=self.rng
        )

        if self.rng.random() < capture_prob:
            return True

        # Failed capture
        pet_agent.apply_failed_capture()
        return False

    def reset_daily(self):
        """Reset daily state (call at start of each day)."""
        self.state.hours_searched_today = 0
        self.state.current_fatigue = max(0, self.state.current_fatigue - 0.8)

    def get_state_snapshot(self) -> Dict[str, Any]:
        """Get snapshot of current state."""
        return {
            "id": self.id,
            "type": self.searcher_type.value,
            "position": self.state.position,
            "is_active": self.state.is_active,
            "fatigue": self.state.current_fatigue,
            "hours_today": self.state.hours_searched_today,
            "total_hours": self.state.total_hours_searched,
        }


def create_search_team(
    num_searchers: int,
    home_position: Tuple[float, float],
    search_radius_m: float,
    include_owner: bool = True,
    seed: Optional[int] = None
) -> List[SearcherAgent]:
    """
    Create a team of searchers with appropriate types and strategies.
    """
    rng = random.Random(seed)
    searchers = []

    for i in range(num_searchers):
        if i == 0 and include_owner:
            # First searcher is always the owner
            searcher = SearcherAgent(
                searcher_id=i,
                searcher_type=SearcherType.OWNER,
                strategy=SearchStrategy.PROFILE_AWARE,
                home_position=home_position,
                search_radius_m=search_radius_m,
                is_owner=True,
                seed=rng.randint(0, 1000000)
            )
        elif i == 1:
            # Second searcher is typically household member
            searcher = SearcherAgent(
                searcher_id=i,
                searcher_type=SearcherType.HOUSEHOLD,
                strategy=SearchStrategy.SIGHTING_CHASER,
                home_position=home_position,
                search_radius_m=search_radius_m,
                is_owner=False,
                seed=rng.randint(0, 1000000)
            )
        else:
            # Additional searchers are volunteers
            searcher = SearcherAgent(
                searcher_id=i,
                searcher_type=SearcherType.VOLUNTEER,
                strategy=rng.choice([
                    SearchStrategy.NAIVE_EXPANDING,
                    SearchStrategy.COORDINATED_GRID,
                    SearchStrategy.SIGHTING_CHASER
                ]),
                home_position=home_position,
                search_radius_m=search_radius_m,
                is_owner=False,
                seed=rng.randint(0, 1000000)
            )

        searchers.append(searcher)

    return searchers
